src/components/Question/Question.js

```diff
@@ -53,9 +53,34 @@ class Question extends Component {
     /* check if current question was already answered correctly */
     const alreadyAnswered =
       score.books[bookId].chapters[chapterId].questions[questionIndex].answered;
+    const alreadyCorrect =
+      score.books[bookId].chapters[chapterId].questions[questionIndex].correct;
 
-    let newScore = score;
-    const newBookScore = score.books.map((book, index) => {
+    /* calculate the numeric score difference 
+     * based on current score state */
+    const scoreDiff = {
+      correct: isCorrect && !alreadyCorrect ? 1 : 0,
+      incorrect: alreadyCorrect
+        ? 0 // previously answered correctly
+        : alreadyAnswered
+          ? isCorrect
+            ? -1 // previously answered, correct answer given
+            : 0 // previously answered, incorrect answer given
+          : isCorrect
+            ? 0 // first attempt, correct answer given
+            : 1, // first attempt, incorrect answer given
+    };
+
+    // return a new score object
+    // NOTE: possible bug; if we declare `newScore` by initializing to
+    // `score`'s properties ( newScore = { books: ..., } etc. )
+    // vs. initializing it as it is below (line 80)
```

Comment:
Also, when implementing the updated score object, I ran into this weird bug(?). My goal was to remove `newScore` altogether and just contain the entire "newScore" object as an argument to `updateScore()` (line 92) without an additional variable declaration. However, this (lines 75 - 79) happens. 

The app behaves as expected with the changes as they are now, but would just like to know why this happens and if I can indeed minify the code a little as I described.

Reply:
Here is a way to update the score immutably:

```javascript
const newScore = {...score, books: score.books.map((book, index) => {
  if (index === bookId) {
    return {
      ...book,
      correct: book.correct + scoreDiff.correct,
      incorrect: book.incorrect + scoreDiff.incorrect,
      chapters: book.chapters.map((chapter, index) => {
        if (index === chapterId) {
          return {
            ...chapter,
            questions: chapter.questions.map((question, index) => {
              if (questionIndex === index) {
                return {
                  answered: true || question.answered,
                  correct: isCorrect,
                };
              }
              return question
            })
          }
        }
        return chapter
      })
    }
  }
  return book
})}
```

In order to use the spread operator we need to add:
`plugins: ['transform-object-rest-spread'],` to the babelrc

Reply:
Just thinking out loud...I feel like the best way to structure `score` would be to not store `book.correct` or `book.incorrect`. Also, for each question, instead of `{answered, correct}`, we just have `{status: "correct" }`, `{status: "incorrect" }`, or `{status: "unanswered" }`

Then we should just use functions to calculate everything we need to get out of the data structure. This doesn't have to be a part of this PR...just a thought I had.